File: 9th_series/implemet stack using one queue/implement_stack_one_queue.cpp
#include <iostream>
#include <queue>
using namespace std;


queue<int> q;

//  push operation
void push(int x) {
    q.push(x);

    int size = q.size();

    // rotate elements so last inserted comes to front
    for (int i = 0; i < size - 1; i++) {
        q.push(q.front());
        q.pop();
    }
}

//  pop operation
int pop() {
    if (q.empty()) {
        cout << "Stack is empty\n";
        return -1;
    }

    int ans = q.front();
    q.pop();
    return ans;
}

//  top operation
int top() {
    if (q.empty()) {
        cout << "Stack is empty\n";
        return -1;
    }

    return q.front();
}

//  empty check
bool isEmpty() {
    return q.empty();
}

//  print function 
void printStack(queue<int> temp) {
    cout << "Stack (top to bottom): ";
    while (!temp.empty()) {
        cout << temp.front() << " ";
        temp.pop();
    }
    cout << endl;
}


int main() {

    push(10);
    push(20);
    push(30);

    printStack(q);   // 30 20 10

    cout << "Top: " << top() << endl;   // 30

    cout << "Pop: " << pop() << endl;   // 30

    printStack(q);   // 20 10

    cout << "Top: " << top() << endl;   // 20

    cout << "Is Empty: " << isEmpty() << endl; // 0 (false)

    pop();
    pop();

    cout << "Is Empty after removing all: " << isEmpty() << endl; // 1 (true)

    return 0;
}

File: 9th_series/implemet stack using one queue/implement_stack_one_queue.js
let q1=[];
// push
function push(x){
    q1.push(x);
}
// pop
function pop(){
    let n=q1.length;
    for(let i=1;i<n;i++){
        q1.push(q1.shift());
    }
    return q1.shift();
}
// top
function top(){
    let n=q1.length;
    for(let i=1;i<n;i++){
        q1.push(q1.shift());
    }
    let ans= q1.shift();
    q1.push(ans);
    return ans;
}
// empty
function empty(){
    return q1.length===0;
}



push(10);
push(20);
push(30);
console.log(q1);
console.log("Top:", top());   // 30
console.log("after top() operation")
console.log(q1);
console.log("Pop:", pop());   // 30
console.log("after pop() operation")
console.log(q1);
console.log("Top:", top());   // 20
console.log("Empty:", empty()); // false